src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -491,9 +497,108 @@ public static String getIndexName(String cfName)
         return StringUtils.substringAfter(cfName, Directories.SECONDARY_INDEX_NAME_SEPARATOR);
     }
 
+    // TODO: JavaDoc if this approach survives...mention that this should not be used for live SSTables
+    public boolean validateStorageAttachedIndexes(Collection<SSTableReader> sstables, boolean failIncomplete) throws IOException
+    {
+        // TODO: A more generic approach via the Index interface rather than just working w/ SAI 
+        Set<StorageAttachedIndex> toValidate = indexes.values().stream().filter(i -> i instanceof StorageAttachedIndex)
+                                                                        .map(i -> (StorageAttachedIndex) i)
+                                                                        .collect(Collectors.toSet());
+
+        if (toValidate.isEmpty())
+            return true;
+        
+        boolean complete = true;
+
+        for (SSTableReader sstable : sstables)
+        {
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+
+            if (indexDescriptor.isPerSSTableIndexBuildComplete())
+            {
+                indexDescriptor.checksumPerSSTableComponents();
```

Comment:
Couldn't this use `indexDescriptor#validatePerSSTableComponents(IndexValidation.CHECKSUM)`, so we don't need the special `IndexDescriptor#checksumPerSSTableComponents` method? And the same for per index components.

Reply:
Those two methods don't preserve the `IOException` throw from the validate methods in `OnDiskFormat`. We could change those methods to throw as well, but then the builder and context manager that call them would have to handle the exception, etc.